File: src/main/java/cc/mirukuneko/realtrainmodrenewed/client/sound/CrossingGateSoundManager.java
package cc.mirukuneko.realtrainmodrenewed.client.sound;

import cc.mirukuneko.realtrainmodrenewed.blockentity.InstalledObjectBlockEntity;
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectDefinition;
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectRegistry;
import net.minecraft.client.Minecraft;
import net.minecraft.client.resources.sounds.AbstractTickableSoundInstance;
import net.minecraft.client.resources.sounds.SoundInstance;
import net.minecraft.core.BlockPos;
import net.minecraft.resources.Identifier;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.level.Level;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class CrossingGateSoundManager {
    private static final Identifier CROSSING_SOUND_ID = Identifier.fromNamespaceAndPath("rtm", "block.crossing_gate");
    private static final Map<String, LoopingCrossingSound> ACTIVE = new ConcurrentHashMap<>();
    /** 踏切音が聞こえる最大距離(ブロック)。これより遠いと鳴らさない/停止する(=どこでも聞こえる不具合対策)。 */
    private static final double MAX_AUDIBLE_DISTANCE = 48.0D;
    private static final double MAX_AUDIBLE_DISTANCE_SQ = MAX_AUDIBLE_DISTANCE * MAX_AUDIBLE_DISTANCE;

    private CrossingGateSoundManager() {
    }

    /** この距離(ブロック)以内は最大音量。これを超えると 0 へ向けて線形フェード。 */
    private static final double FULL_VOLUME_DISTANCE = 12.0D;

    /** クライアントプレイヤーが pos から可聴距離内か。プレイヤー不在時は false(鳴らさない)。 */
    private static boolean playerInRange(BlockPos pos) {
        net.minecraft.client.player.LocalPlayer player = Minecraft.getInstance().player;
        if (player == null) {
            return false;
        }
        double dx = (pos.getX() + 0.5D) - player.getX();
        double dy = (pos.getY() + 0.5D) - player.getY();
        double dz = (pos.getZ() + 0.5D) - player.getZ();
        return (dx * dx + dy * dy + dz * dz) <= MAX_AUDIBLE_DISTANCE_SQ;
    }

    /** プレイヤーとの距離に応じた音量(近=1.0、FULL_VOLUME_DISTANCE超で線形減衰、MAXで0)。 */
    private static float volumeForDistance(BlockPos pos) {
        net.minecraft.client.player.LocalPlayer player = Minecraft.getInstance().player;
        if (player == null) {
            return 0.0F;
        }
        double dx = (pos.getX() + 0.5D) - player.getX();
        double dy = (pos.getY() + 0.5D) - player.getY();
        double dz = (pos.getZ() + 0.5D) - player.getZ();
        double dist = Math.sqrt(dx * dx + dy * dy + dz * dz);
        if (dist <= FULL_VOLUME_DISTANCE) {
            return 1.0F;
        }
        if (dist >= MAX_AUDIBLE_DISTANCE) {
            return 0.0F;
        }
        double t = (dist - FULL_VOLUME_DISTANCE) / (MAX_AUDIBLE_DISTANCE - FULL_VOLUME_DISTANCE);
        return (float) (1.0D - t);
    }

    public static void tick(InstalledObjectBlockEntity blockEntity) {
        if (blockEntity == null) {
            return;
        }
        Level level = blockEntity.getLevel();
        if (level == null || !level.isClientSide()) {
            return;
        }
        if (!blockEntity.isPowered()) {
            stop(level, blockEntity.getBlockPos());
            return;
        }
        if (!shouldPlayCrossingSound(blockEntity)) {
            stop(level, blockEntity.getBlockPos());
            return;
        }
        // 可聴距離外なら鳴らさない/停止(=どこでも聞こえる不具合対策)。
        if (!playerInRange(blockEntity.getBlockPos())) {
            stop(level, blockEntity.getBlockPos());
            return;
        }

        Minecraft minecraft = Minecraft.getInstance();
        if (minecraft.getSoundManager() == null) {
            return;
        }

        String key = key(level, blockEntity.getBlockPos());
        Identifier soundId = resolveSoundId(blockEntity);
        LoopingCrossingSound sound = ACTIVE.get(key);
        if (sound == null || sound.isStopped() || !sound.matches(soundId)) {
            if (sound != null) {
                sound.requestStop();
            }
            sound = new LoopingCrossingSound(blockEntity, soundId);
            ACTIVE.put(key, sound);
            minecraft.getSoundManager().play(sound);
        } else {
            sound.refresh();
        }
    }

    public static void stop(Level level, BlockPos pos) {
        if (level == null || pos == null) {
            return;
        }
        LoopingCrossingSound sound = ACTIVE.remove(key(level, pos));
        if (sound != null) {
            sound.requestStop();
        }
    }

    private static String key(Level level, BlockPos pos) {
        return level.dimension().identifier() + "|" + pos.asLong();
    }

    private static boolean shouldPlayCrossingSound(InstalledObjectBlockEntity blockEntity) {
        InstalledObjectDefinition definition = blockEntity == null ? null : InstalledObjectRegistry.getById(blockEntity.getDefinitionId());
        if (definition == null) {
            return false;
        }
        String runningSound = definition.getRunningSound();
        return runningSound != null && !runningSound.isBlank();
    }

    private static Identifier resolveSoundId(InstalledObjectBlockEntity blockEntity) {
        InstalledObjectDefinition definition = blockEntity == null ? null : InstalledObjectRegistry.getById(blockEntity.getDefinitionId());
        String raw = definition == null ? "" : definition.getRunningSound();
        if (raw == null || raw.isBlank()) {
            return CROSSING_SOUND_ID;
        }
        String normalized = raw.trim().replace('\\', '/');
        if (normalized.endsWith(".ogg")) {
            normalized = normalized.substring(0, normalized.length() - 4);
        }
        if (normalized.startsWith("sounds/")) {
            normalized = normalized.substring("sounds/".length());
        }
        String lowered = normalized.toLowerCase(java.util.Locale.ROOT);
        if (lowered.contains("rtm_crossinggate") || lowered.contains("crossinggate0") || lowered.contains("crossinggate1")) {
            return CROSSING_SOUND_ID;
        }
        try {
            if (normalized.contains(":")) {
                String[] split = normalized.split(":", 2);
                String namespace = split[0].isBlank() ? "minecraft" : split[0].toLowerCase(java.util.Locale.ROOT);
                String path = split[1].toLowerCase(java.util.Locale.ROOT);
                if ("rtm".equals(namespace) && path.indexOf('/') >= 0) {
                    path = path.replace('/', '.');
                }
                return Identifier.fromNamespaceAndPath(namespace, path);
            }
            String path = normalized.toLowerCase(java.util.Locale.ROOT);
            if (path.indexOf('/') >= 0) {
                path = path.replace('/', '.');
            }
            return Identifier.fromNamespaceAndPath("rtm", path);
        } catch (Exception ignored) {
            return CROSSING_SOUND_ID;
        }
    }

    private static final class LoopingCrossingSound extends AbstractTickableSoundInstance {
        private final InstalledObjectBlockEntity blockEntity;
        private final Identifier soundId;

        private LoopingCrossingSound(InstalledObjectBlockEntity blockEntity, Identifier soundId) {
            super(SoundEvent.createVariableRangeEvent(soundId == null ? CROSSING_SOUND_ID : soundId), SoundSource.BLOCKS, SoundInstance.createUnseededRandom());
            this.blockEntity = blockEntity;
            this.soundId = soundId == null ? CROSSING_SOUND_ID : soundId;
            this.looping = true;
            this.delay = 0;
            this.relative = false;
            // 距離減衰は MC 任せにせず手動で音量を調整する(可聴端で急に消えるのを防ぎ徐々にフェード)。
            this.attenuation = SoundInstance.Attenuation.NONE;
            this.volume = 1.0F;
            this.pitch = 1.0F;
            refresh();
        }

        private boolean matches(Identifier other) {
            return soundId.equals(other == null ? CROSSING_SOUND_ID : other);
        }

        private void refresh() {
            BlockPos pos = blockEntity.getBlockPos();
            this.x = pos.getX() + 0.5D;
            this.y = pos.getY() + 0.5D;
            this.z = pos.getZ() + 0.5D;
            // プレイヤーとの距離で音量を滑らかに変える(近=最大、可聴端48mで0)。
            this.volume = volumeForDistance(pos);
        }

        private void requestStop() {
            stop();
        }

        @Override
        public void tick() {
            Level level = blockEntity.getLevel();
            if (level == null
                || !level.isClientSide()
                || blockEntity.isRemoved()
                || !blockEntity.isPowered()
                || !shouldPlayCrossingSound(blockEntity)
                || !playerInRange(blockEntity.getBlockPos())
                || level.getBlockEntity(blockEntity.getBlockPos()) != blockEntity) {
                if (level != null) {
                    ACTIVE.remove(key(level, blockEntity.getBlockPos()), this);
                }
                stop();
                return;
            }
            refresh();
        }
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/client/sound/CrossingGateSoundManager.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.client.sound

import cc.mirukuneko.realtrainmodrenewed.blockentity.InstalledObjectBlockEntity
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectDefinition
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectRegistry
import net.minecraft.client.Minecraft
import net.minecraft.client.resources.sounds.AbstractTickableSoundInstance
import net.minecraft.client.resources.sounds.SoundInstance
import net.minecraft.core.BlockPos
import net.minecraft.resources.Identifier
import net.minecraft.sounds.SoundEvent
import net.minecraft.sounds.SoundSource
import net.minecraft.world.level.Level
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap

object CrossingGateSoundManager {
    private val CROSSING_SOUND_ID: Identifier = Identifier.fromNamespaceAndPath("rtm", "block.crossing_gate")
    private val ACTIVE: MutableMap<String, LoopingCrossingSound> = ConcurrentHashMap()
    private const val MAX_AUDIBLE_DISTANCE = 48.0
    private const val MAX_AUDIBLE_DISTANCE_SQ = MAX_AUDIBLE_DISTANCE * MAX_AUDIBLE_DISTANCE
    private const val FULL_VOLUME_DISTANCE = 12.0

    private fun playerInRange(pos: BlockPos): Boolean {
        val player = Minecraft.getInstance().player ?: return false
        val dx = pos.x + 0.5 - player.x
        val dy = pos.y + 0.5 - player.y
        val dz = pos.z + 0.5 - player.z
        return dx * dx + dy * dy + dz * dz <= MAX_AUDIBLE_DISTANCE_SQ
    }

    private fun volumeForDistance(pos: BlockPos): Float {
        val player = Minecraft.getInstance().player ?: return 0.0f
        val dx = pos.x + 0.5 - player.x
        val dy = pos.y + 0.5 - player.y
        val dz = pos.z + 0.5 - player.z
        val dist = kotlin.math.sqrt(dx * dx + dy * dy + dz * dz)
        if (dist <= FULL_VOLUME_DISTANCE) return 1.0f
        if (dist >= MAX_AUDIBLE_DISTANCE) return 0.0f
        val t = (dist - FULL_VOLUME_DISTANCE) / (MAX_AUDIBLE_DISTANCE - FULL_VOLUME_DISTANCE)
        return (1.0 - t).toFloat()
    }

    @JvmStatic
    fun tick(blockEntity: InstalledObjectBlockEntity?) {
        if (blockEntity == null) return
        val level = blockEntity.level
        if (level == null || !level.isClientSide) return
        if (!blockEntity.isPowered) {
            stop(level, blockEntity.blockPos)
            return
        }
        if (!shouldPlayCrossingSound(blockEntity)) {
            stop(level, blockEntity.blockPos)
            return
        }
        if (!playerInRange(blockEntity.blockPos)) {
            stop(level, blockEntity.blockPos)
            return
        }

        val minecraft = Minecraft.getInstance()
        val key = key(level, blockEntity.blockPos)
        val soundId = resolveSoundId(blockEntity)
        var sound = ACTIVE[key]
        if (sound == null || sound.isStopped || !sound.matches(soundId)) {
            sound?.requestStop()
            sound = LoopingCrossingSound(blockEntity, soundId)
            ACTIVE[key] = sound
            minecraft.soundManager.play(sound)
        } else {
            sound.refresh()
        }
    }

    @JvmStatic
    fun stop(level: Level?, pos: BlockPos?) {
        if (level == null || pos == null) return
        ACTIVE.remove(key(level, pos))?.requestStop()
    }

    private fun key(level: Level, pos: BlockPos): String =
        "${level.dimension().identifier()}|${pos.asLong()}"

    private fun shouldPlayCrossingSound(blockEntity: InstalledObjectBlockEntity?): Boolean {
        val definition = blockEntity?.let { InstalledObjectRegistry.getById(it.definitionId) }
        val runningSound = definition?.runningSound
        return !runningSound.isNullOrBlank()
    }

    private fun resolveSoundId(blockEntity: InstalledObjectBlockEntity?): Identifier {
        val definition: InstalledObjectDefinition? = blockEntity?.let { InstalledObjectRegistry.getById(it.definitionId) }
        val raw = definition?.runningSound
        if (raw.isNullOrBlank()) return CROSSING_SOUND_ID
        var normalized = raw.trim().replace('\\', '/')
        if (normalized.endsWith(".ogg")) {
            normalized = normalized.substring(0, normalized.length - 4)
        }
        if (normalized.startsWith("sounds/")) {
            normalized = normalized.substring("sounds/".length)
        }
        val lowered = normalized.lowercase(Locale.ROOT)
        if (lowered.contains("rtm_crossinggate") || lowered.contains("crossinggate0") || lowered.contains("crossinggate1")) {
            return CROSSING_SOUND_ID
        }
        return try {
            if (normalized.contains(":")) {
                val split = normalized.split(":", limit = 2)
                val namespace = split[0].ifBlank { "minecraft" }.lowercase(Locale.ROOT)
                var path = split[1].lowercase(Locale.ROOT)
                if (namespace == "rtm" && path.indexOf('/') >= 0) {
                    path = path.replace('/', '.')
                }
                Identifier.fromNamespaceAndPath(namespace, path)
            } else {
                var path = normalized.lowercase(Locale.ROOT)
                if (path.indexOf('/') >= 0) {
                    path = path.replace('/', '.')
                }
                Identifier.fromNamespaceAndPath("rtm", path)
            }
        } catch (_: Exception) {
            CROSSING_SOUND_ID
        }
    }

    private class LoopingCrossingSound(
        private val blockEntity: InstalledObjectBlockEntity,
        soundId: Identifier?
    ) : AbstractTickableSoundInstance(
        SoundEvent.createVariableRangeEvent(soundId ?: CROSSING_SOUND_ID),
        SoundSource.BLOCKS,
        SoundInstance.createUnseededRandom()
    ) {
        private val soundId: Identifier = soundId ?: CROSSING_SOUND_ID

        init {
            looping = true
            delay = 0
            relative = false
            attenuation = SoundInstance.Attenuation.NONE
            volume = 1.0f
            pitch = 1.0f
            refresh()
        }

        fun matches(other: Identifier?): Boolean =
            soundId == (other ?: CROSSING_SOUND_ID)

        fun refresh() {
            val pos = blockEntity.blockPos
            x = pos.x + 0.5
            y = pos.y + 0.5
            z = pos.z + 0.5
            volume = volumeForDistance(pos)
        }

        fun requestStop() {
            stop()
        }

        override fun tick() {
            val level = blockEntity.level
            if (level == null ||
                !level.isClientSide ||
                blockEntity.isRemoved ||
                !blockEntity.isPowered ||
                !shouldPlayCrossingSound(blockEntity) ||
                !playerInRange(blockEntity.blockPos) ||
                level.getBlockEntity(blockEntity.blockPos) !== blockEntity
            ) {
                if (level != null) {
                    ACTIVE.remove(key(level, blockEntity.blockPos), this)
                }
                stop()
                return
            }
            refresh()
        }
    }
}
